snowdrop-core/src/Snowdrop/Core/Validator/Types.hs

```diff
@@ -1,89 +1,122 @@
+{-# LANGUAGE UndecidableSuperClasses #-}
+{-# LANGUAGE AllowAmbiguousTypes #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE DataKinds #-}
+
 module Snowdrop.Core.Validator.Types
        ( PreValidator (..)
        , mkPreValidator
+       , unitePreValidators
        , Validator
        , mkValidator
        , runValidator
        , upcastPreValidator
-       , downcastPreValidator
-       , castPreValidator
+       -- , castPreValidator
        , fromPreValidator
        , getPreValidator
        ) where
 
-import           Universum
+import           Universum hiding (Nat)
 
-import           Control.Monad.Except (throwError)
 import           Data.Vinyl
+import           Data.Vinyl.TypeLevel (Nat (..), type RLength)
+
+import           Snowdrop.Core.ERoComp (ERoComp, UpCastableERo, upcastEffERoComp)
+import           Snowdrop.Core.Transaction (DownCastableTx, StateTx, TxComponents, downcastStateTx)
+import           Snowdrop.Core.Stuff (UnitedTxType)
 
-import           Snowdrop.Core.ERoComp.Types
-import           Snowdrop.Core.Transaction
+import           Snowdrop.Util (RContains, RecAll', RecToList (..))
 
-import           Snowdrop.Util (HasGetter, HasPrism (..), RContains)
+------------------------------------------------------
+-- PreValidator
+------------------------------------------------------
 
--- | Prevalidator validates part of the transaction.
-newtype PreValidator e id value ctx txtype =
-    PreValidator { runPrevalidator :: StateTx id value txtype -> ERoComp e id value ctx () }
+-- pva701: maybe we should create ValidatorComponents and use it instead of TxComponents
```

Comment:
Add `TODO` into the comment text pls

Reply:
What's the difference BTW except for naming?

Reply:
you may want to get access to some extra components (apart from `TxComponents`)... don't have any real examples though, but they can appear in future.